File: Sharebook/Server/src/Services/Post/Sharebook.Post.API/Parameters/TimelineRequestParameters.cs
namespace Sharebook.Post.Application.Parameters
{
    using Sharebook.Common.Public.Parameters;
    using Sharebook.Common.Settings;

    public interface ITimelineRequestParameters : IMinMaxQueryParameters
    {
        /// <summary>
        /// If set to true, the creator property (IUser) will only contain the id.
        /// </summary>
        bool? TrimUser { get; set; }

        /// <summary>
        /// Include tweet entities.
        /// </summary>
        bool? IncludeEntities { get; set; }

        /// <summary>
        /// Decide whether to use Extended or Compat mode
        /// </summary>
        TweetMode? TweetMode { get; set; }
    }

    public abstract class TimelineRequestParameters : MinMaxQueryParameters, ITimelineRequestParameters
    {
        protected TimelineRequestParameters()
        {
        }

        protected TimelineRequestParameters(ITimelineRequestParameters source) : base(source)
        {
            this.TrimUser = source?.TrimUser;
            this.IncludeEntities = source?.IncludeEntities;
        }

        /// <inheritdoc/>
        public bool? TrimUser { get; set; }

        /// <inheritdoc/>
        public bool? IncludeEntities { get; set; }

        /// <inheritdoc/>
        public TweetMode? TweetMode { get; set; }
    }
}

File: Sharebook/Client/src/app/core/Public/Parameters/TimelineRequestParameters.ts
import {IMinMaxQueryParameters, MinMaxQueryParameters} from "./MaxAndMinBaseQueryParameters";
import {ITweetModeParameter} from "./ITweetModeParameter";
import {TweetMode} from '../Settings/SharebookSettings';

export interface ITimelineRequestParameters extends IMinMaxQueryParameters, ITweetModeParameter {
  // If set to true, the creator property (IUser) will only contain the id.
  trimUser: boolean | null;

  // Include tweet entities.
  includeEntities: boolean | null;
}

export abstract class TimelineRequestParameters extends MinMaxQueryParameters implements ITimelineRequestParameters {
  protected constructor(source?: ITimelineRequestParameters) {
    if (source) {
      super(source);

      this.trimUser = source?.trimUser;
      this.includeEntities = source?.includeEntities;
      this.tweetMode = source?.tweetMode;
    } else {
      super();
    }
  }

  public trimUser: boolean | null;
  public includeEntities: boolean | null;
  public tweetMode: TweetMode | null;
}

// protected TimelineRequestParameters()
// {
// }
//
// protected TimelineRequestParameters(ITimelineRequestParameters source) : base(source)
// {
//   TrimUser = source?.TrimUser;
//   IncludeEntities = source?.IncludeEntities;
//   TweetMode = source?.TweetMode;
// }
